### ISSUE TITLE
SSZ map concurrent read-write

### ISSUE THREAD

**Issue by meyer9** [prysmaticlabs/prysm#2671]
Really not sure why this is happening because it's protected by a lock, but I'm getting a concurrent map read/map write error here:

```
fatal error: concurrent map read and map write

goroutine 380 [running]:
runtime.throw(0x1266d9e, 0x21)
	/usr/lib/go/src/runtime/panic.go:617 +0x72 fp=0xc00a5207c8 sp=0xc00a520798 pc=0x433692
runtime.mapaccess1(0x1104d60, 0xc0002b4000, 0xc00a5208b0, 0x8)
	/usr/lib/go/src/runtime/map.go:413 +0x387 fp=0xc00a520848 sp=0xc00a5207c8 pc=0x4101c7
github.com/prysmaticlabs/prysm/shared/ssz.cachedSSZUtilsNoAcquireLock(0x14df520, 0x10c8d60, 0x0, 0x0, 0x0)
	/home/meyer9/go/pkg/mod/github.com/prysmaticlabs/prysm@v0.0.0-20190514144052-d672a0602655/shared/ssz/ssz_utils_cache.go:59 +0x7b fp=0xc00a5208d0 sp=0xc00a520848 pc=0xf0e6db
github.com/prysmaticlabs/prysm/shared/ssz.getEncoding(0x10c8d60, 0xc008f62ea0, 0x18b, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/meyer9/go/pkg/mod/github.com/prysmaticlabs/prysm@v0.0.0-20190514144052-d672a0602655/shared/ssz/hash.go:88 +0xc5 fp=0xc00a520988 sp=0xc00a5208d0 pc=0xf0a975
github.com/prysmaticlabs/prysm/shared/ssz.makeStructHasher.func1(0x11fc400, 0xc008f62e00, 0x199, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/meyer9/go/pkg/mod/github.com/prysmaticlabs/prysm@v0.0.0-20190514144052-d672a0602655/shared/ssz/hash.go:143 +0x201 fp=0xc00a520b50 sp=0xc00a520988 pc=0xf13a01
github.com/prysmaticlabs/prysm/shared/ssz.makeSliceHasher.func1(0x10b2a80, 0xc00a4f4f30, 0x97, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/meyer9/go/pkg/mod/github.com/prysmaticlabs/prysm@v0.0.0-20190514144052-d672a0602655/shared/ssz/hash.go:120 +0x174 fp=0xc00a520ce8 sp=0xc00a520b50 pc=0xf132c4
github.com/prysmaticlabs/prysm/shared/ssz.makeStructHasher.func1(0x1234460, 0xc00a4f4f00, 0x99, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/meyer9/go/pkg/mod/github.com/prysmaticlabs/prysm@v0.0.0-20190514144052-d672a0602655/shared/ssz/hash.go:143 +0x201 fp=0xc00a520eb0 sp=0xc00a520ce8 pc=0xf13a01
github.com/prysmaticlabs/prysm/shared/ssz.TreeHash(0x1234460, 0xc00a4f4f00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/meyer9/go/pkg/mod/github.com/prysmaticlabs/prysm@v0.0.0-20190514144052-d672a0602655/shared/ssz/hash.go:34 +0x36a fp=0xc00a521040 sp=0xc00a520eb0 pc=0xf09aca
github.com/phoreproject/synapse/beacon/rpc.(*server).GetStateRoot(0xc00a4dc020, 0x14cde00, 0xc00a4e9170, 0xc00a4dcf00, 0x0, 0x0, 0x0)
	/home/meyer9/development/synapse/beacon/rpc/rpc.go:144 +0x117 fp=0xc00a521558 sp=0xc00a521040 pc=0x101b987
github.com/phoreproject/synapse/pb._BlockchainRPC_GetStateRoot_Handler(0x11e7c40, 0xc00a4dc020, 0x14cde00, 0xc00a4e9170, 0xc00b8f4960, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/meyer9/development/synapse/pb/rpc.pb.go:1258 +0x3df fp=0xc00a521668 sp=0xc00a521558 pc=0xca497f
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00a4f4000, 0x14d6ae0, 0xc00b8f2300, 0xc00bbe4100, 0xc00a4e8180, 0x1cd7bb8, 0x0, 0x0, 0x0)
	/home/meyer9/go/pkg/mod/google.golang.org/grpc@v1.20.1/server.go:972 +0x1038 fp=0xc00a521c10 sp=0xc00a521668 pc=0xc72198
google.golang.org/grpc.(*Server).handleStream(0xc00a4f4000, 0x14d6ae0, 0xc00b8f2300, 0xc00bbe4100, 0x0)
	/home/meyer9/go/pkg/mod/google.golang.org/grpc@v1.20.1/server.go:1252 +0x976 fp=0xc00a521f70 sp=0xc00a521c10 pc=0xc76206
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc00b8b05c0, 0xc00a4f4000, 0x14d6ae0, 0xc00b8f2300, 0xc00bbe4100)
	/home/meyer9/go/pkg/mod/google.golang.org/grpc@v1.20.1/server.go:691 +0xb5 fp=0xc00a521fb8 sp=0xc00a521f70 pc=0xc869e5
runtime.goexit()
	/usr/lib/go/src/runtime/asm_amd64.s:1337 +0x1 fp=0xc00a521fc0 sp=0xc00a521fb8 pc=0x462af1
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/home/meyer9/go/pkg/mod/google.golang.org/grpc@v1.20.1/server.go:689 +0xb0
```

**Comment by meyer9** [comment]
Run with `-race`:
```
==================
WARNING: DATA RACE
Read at 0x00c000314000 by goroutine 117:
  runtime.mapaccess1()
      /usr/lib/go/src/runtime/map.go:396 +0x0
  github.com/prysmaticlabs/prysm/shared/ssz.cachedSSZUtilsNoAcquireLock()
      /home/meyer9/go/pkg/mod/github.com/prysmaticlabs/prysm@v0.0.0-20190514144052-d672a0602655/shared/ssz/ssz_utils_cache.go:59 +0x98
  github.com/prysmaticlabs/prysm/shared/ssz.getEncoding()
      /home/meyer9/go/pkg/mod/github.com/prysmaticlabs/prysm@v0.0.0-20190514144052-d672a0602655/shared/ssz/hash.go:88 +0x82
  github.com/prysmaticlabs/prysm/shared/ssz.hashedEncoding()
      /home/meyer9/go/pkg/mod/github.com/prysmaticlabs/prysm@v0.0.0-20190514144052-d672a0602655/shared/ssz/hash.go:104 +0x6a
  github.com/prysmaticlabs/prysm/shared/ssz.makeStructHasher.func1()
      /home/meyer9/go/pkg/mod/github.com/prysmaticlabs/prysm@v0.0.0-20190514144052-d672a0602655/shared/ssz/hash.go:143 +0x1a0
  github.com/prysmaticlabs/prysm/shared/ssz.makeSliceHasher.func1()
      /home/meyer9/go/pkg/mod/github.com/prysmaticlabs/prysm@v0.0.0-20190514144052-d672a0602655/shared/ssz/hash.go:120 +0x17f
  github.com/prysmaticlabs/prysm/shared/ssz.makeStructHasher.func1()
      /home/meyer9/go/pkg/mod/github.com/prysmaticlabs/prysm@v0.0.0-20190514144052-d672a0602655/shared/ssz/hash.go:143 +0x1a0
  github.com/prysmaticlabs/prysm/shared/ssz.TreeHash()
      /home/meyer9/go/pkg/mod/github.com/prysmaticlabs/prysm@v0.0.0-20190514144052-d672a0602655/shared/ssz/hash.go:34 +0x324
  github.com/phoreproject/synapse/beacon/rpc.(*server).GetStateRoot()
      /home/meyer9/development/synapse/beacon/rpc/rpc.go:144 +0x178
  github.com/phoreproject/synapse/pb._BlockchainRPC_GetStateRoot_Handler()
      /home/meyer9/development/synapse/pb/rpc.pb.go:1258 +0x2fc
  google.golang.org/grpc.(*Server).processUnaryRPC()
      /home/meyer9/go/pkg/mod/google.golang.org/grpc@v1.20.1/server.go:972 +0x936
  google.golang.org/grpc.(*Server).handleStream()
      /home/meyer9/go/pkg/mod/google.golang.org/grpc@v1.20.1/server.go:1252 +0x1351
  google.golang.org/grpc.(*Server).serveStreams.func1.1()
      /home/meyer9/go/pkg/mod/google.golang.org/grpc@v1.20.1/server.go:691 +0xac

Previous write at 0x00c000314000 by goroutine 118:
  runtime.mapassign()
      /usr/lib/go/src/runtime/map.go:576 +0x0
  github.com/prysmaticlabs/prysm/shared/ssz.cachedSSZUtilsNoAcquireLock()
      /home/meyer9/go/pkg/mod/github.com/prysmaticlabs/prysm@v0.0.0-20190514144052-d672a0602655/shared/ssz/ssz_utils_cache.go:66 +0x135
  github.com/prysmaticlabs/prysm/shared/ssz.cachedSSZUtils()
      /home/meyer9/go/pkg/mod/github.com/prysmaticlabs/prysm@v0.0.0-20190514144052-d672a0602655/shared/ssz/ssz_utils_cache.go:48 +0x127
  github.com/prysmaticlabs/prysm/shared/ssz.TreeHash()
      /home/meyer9/go/pkg/mod/github.com/prysmaticlabs/prysm@v0.0.0-20190514144052-d672a0602655/shared/ssz/hash.go:30 +0xfd
  github.com/phoreproject/synapse/primitives.(*State).ValidateAttestation()
      /home/meyer9/development/synapse/primitives/state.go:1747 +0xa85
  github.com/phoreproject/synapse/beacon.(*Mempool).ProcessNewAttestation()
      /home/meyer9/development/synapse/beacon/mempool.go:68 +0x402
  github.com/phoreproject/synapse/beacon/rpc.(*server).SubmitAttestation()
      /home/meyer9/development/synapse/beacon/rpc/rpc.go:42 +0x12a
  github.com/phoreproject/synapse/pb._BlockchainRPC_SubmitAttestation_Handler()
      /home/meyer9/development/synapse/pb/rpc.pb.go:1348 +0x2fc
  google.golang.org/grpc.(*Server).processUnaryRPC()
      /home/meyer9/go/pkg/mod/google.golang.org/grpc@v1.20.1/server.go:972 +0x936
  google.golang.org/grpc.(*Server).handleStream()
      /home/meyer9/go/pkg/mod/google.golang.org/grpc@v1.20.1/server.go:1252 +0x1351
  google.golang.org/grpc.(*Server).serveStreams.func1.1()
      /home/meyer9/go/pkg/mod/google.golang.org/grpc@v1.20.1/server.go:691 +0xac

Goroutine 117 (running) created at:
  google.golang.org/grpc.(*Server).serveStreams.func1()
      /home/meyer9/go/pkg/mod/google.golang.org/grpc@v1.20.1/server.go:689 +0xb8
  google.golang.org/grpc/internal/transport.(*http2Server).operateHeaders()
      /home/meyer9/go/pkg/mod/google.golang.org/grpc@v1.20.1/internal/transport/http2_server.go:421 +0x1622
  google.golang.org/grpc/internal/transport.(*http2Server).HandleStreams()
      /home/meyer9/go/pkg/mod/google.golang.org/grpc@v1.20.1/internal/transport/http2_server.go:461 +0x37a
  google.golang.org/grpc.(*Server).serveStreams()
      /home/meyer9/go/pkg/mod/google.golang.org/grpc@v1.20.1/server.go:687 +0x170
  google.golang.org/grpc.(*Server).handleRawConn.func1()
      /home/meyer9/go/pkg/mod/google.golang.org/grpc@v1.20.1/server.go:649 +0x50

Goroutine 118 (running) created at:
  google.golang.org/grpc.(*Server).serveStreams.func1()
      /home/meyer9/go/pkg/mod/google.golang.org/grpc@v1.20.1/server.go:689 +0xb8
  google.golang.org/grpc/internal/transport.(*http2Server).operateHeaders()
      /home/meyer9/go/pkg/mod/google.golang.org/grpc@v1.20.1/internal/transport/http2_server.go:421 +0x1622
  google.golang.org/grpc/internal/transport.(*http2Server).HandleStreams()
      /home/meyer9/go/pkg/mod/google.golang.org/grpc@v1.20.1/internal/transport/http2_server.go:461 +0x37a
  google.golang.org/grpc.(*Server).serveStreams()
      /home/meyer9/go/pkg/mod/google.golang.org/grpc@v1.20.1/server.go:687 +0x170
  google.golang.org/grpc.(*Server).handleRawConn.func1()
      /home/meyer9/go/pkg/mod/google.golang.org/grpc@v1.20.1/server.go:649 +0x50
==================
```

**Comment by meyer9** [comment]
Seems that this code path doesn't acquire the lock:
```
  github.com/prysmaticlabs/prysm/shared/ssz.cachedSSZUtilsNoAcquireLock()
      /home/meyer9/go/pkg/mod/github.com/prysmaticlabs/prysm@v0.0.0-20190514144052-d672a0602655/shared/ssz/ssz_utils_cache.go:59 +0x98
  github.com/prysmaticlabs/prysm/shared/ssz.getEncoding()
      /home/meyer9/go/pkg/mod/github.com/prysmaticlabs/prysm@v0.0.0-20190514144052-d672a0602655/shared/ssz/hash.go:88 +0x82
  github.com/prysmaticlabs/prysm/shared/ssz.hashedEncoding()
      /home/meyer9/go/pkg/mod/github.com/prysmaticlabs/prysm@v0.0.0-20190514144052-d672a0602655/shared/ssz/hash.go:104 +0x6a
  github.com/prysmaticlabs/prysm/shared/ssz.makeStructHasher.func1()
      /home/meyer9/go/pkg/mod/github.com/prysmaticlabs/prysm@v0.0.0-20190514144052-d672a0602655/shared/ssz/hash.go:143 +0x1a0
  github.com/prysmaticlabs/prysm/shared/ssz.makeSliceHasher.func1()
      /home/meyer9/go/pkg/mod/github.com/prysmaticlabs/prysm@v0.0.0-20190514144052-d672a0602655/shared/ssz/hash.go:120 +0x17f
  github.com/prysmaticlabs/prysm/shared/ssz.makeStructHasher.func1()
      /home/meyer9/go/pkg/mod/github.com/prysmaticlabs/prysm@v0.0.0-20190514144052-d672a0602655/shared/ssz/hash.go:143 +0x1a0
  github.com/prysmaticlabs/prysm/shared/ssz.TreeHash()
```

**Comment by meyer9** [comment]
Reproduced here: https://github.com/meyer9/prysm/tree/ssz-data-race

**Comment by meyer9** [comment]
Fixed with a94e73b2d6d20fc851355d1c6bec52ab9318986f.
